### PR TITLE
Fix #3333: adapt child instantiation

### LINKED ISSUE #3333: Patmat exhaustivity checks do not consider all possible type children
There are (at least) two cases in which children of a type are not considered in patmat exhaustivity checks. I'm making a single issue, since it seems that the solution will have to handle both of these problems at once.

When compiling the following code:
```
sealed trait IntegralNumber

object IntegralNumber {
  object NaN extends IntegralNumber
  object NotNaN extends IntegralNumber

  sealed abstract class FiniteNumberImpl[N](val value: N) extends IntegralNumber
  sealed class IntNumber(value: Int) extends FiniteNumberImpl[Int](value)

  def test(o: IntegralNumber) = o match {
    case NaN => -1
  }

}
```
the following warning is emitted:
```
10 |  def test(o: IntegralNumber) = o match {
   |                                ^
   |                                match may not be exhaustive.
   |
   |                                It would fail on: NotNaN
```
Note that `IntNumber` is valid input to `test`. This happens, since `SpaceEngine#refine` initializes `FiniteNumberImp[N]` to `FiniteNumberImpl[Any]`, and then (correctly) when `FiniteNumberImpl[Any]` is decomposed, `IntNumber` is ousted.

On a somewhat related note, traits also are ousted too eagerly when refined types are involved. When compiling the following code:
```
sealed trait T
object O1 extends T
object O2 extends T { val x: Int = 0 }

sealed trait TT extends T
object O3 extends TT { val x: Int = 0 }

sealed trait Eliminated extends T

object O {
  def m(t: T { val x: Int }) = t match {
    case _: Eliminated => ;
  }
}
```
the following warning is emitted:
```
11 |  def m(t: T { val x: Int }) = t match {
   |                               ^
   |                               match may not be exhaustive.
   |
   |                               It would fail on: O2
```
This happens, since `SpaceEngine#refine` cannot `instantiate` `TT` such that it is a subtype of `T { val x: Int }` and removes it, even though its children can (and do) fulfill the refinement.

@liufengyun - this is as far as I've gotten with this issue. I tried to adjust `instantiate` to handle these cases correctly, but that seems a bit too difficult for me now.

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #3333 #3455 #3469: adapt child instantiation

Structural types in #3333 is left open.

- If a child type parameter is not constrained, instantiate it to Wildcard
- In prefix inference, forget `ThisType` for modules.

### CODE REVIEW

**Comment by liufengyun** [comment]
@AleksanderBG Do you want to review? This may be helpful to your WIP.

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 1]
```diff
@@ -611,13 +609,32 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
...
+      minimizeAll = false
+    )
+
     val tvars = tp1.typeParams.map { tparam => newTypeVar(tparam.paramInfo.bounds) }
     val protoTp1 = thisTypeMap(tp1.appliedTo(tvars))
 
-    if (protoTp1 <:< tp2 && isFullyDefined(protoTp1, ForceDegree.noBottom)) protoTp1
+    if (protoTp1 <:< tp2) {
+      isFullyDefined(protoTp1, force)
+      instUndetMap(protoTp1)
```
Comment (AleksanderBG) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: If I understand correctly how `isFullyDefined` behaves, then if it returns `true`, all TVars in `protoTp1` should be already instantiated and `instUndetMap` will be a no-op. Would it make sense to only use `instUndetMap` if `isFullyDefined` returned `false`?

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 2]
```diff
@@ -584,10 +584,9 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
     // precondition: `tp1` should have the shape `path.Child`, thus `ThisType` is always covariant
     val thisTypeMap = new TypeMap {
       def apply(t: Type): Type = t match {
-        case tp @ ThisType(tref) if !tref.symbol.isStaticOwner && !tref.symbol.is(Module)  =>
-          // TODO: stackoverflow here
-          // newTypeVar(TypeBounds.upper(mapOver(tp.underlying)))
```
[the remark this answers is not in the document]
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: I forgot exactly what's the problem, but it's related to `ClassInfo#selfType`. `tp.underlying` is supposed to be decreasing (I assume), so I think the case you mention will not occur. For the task you want to deal with, the memoization in your code is correct.

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala (hunk 1 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Yes, good catch, I've updated the code, thanks a lot.

**Pull request merged**